### DISCUSSION TITLE
Testing errors

### DISCUSSION THREAD

**Discussion by nazardiachok** [neuefische/questions#318] [Web] [answered]
Hello everybody! Can anybody help me please to solve the errors in testing?  I have two tests: 1) testing of the input fields (PersonalData.js > PersonalData.test.js), 2) testing of the function (Cards.js > Cards.test.js) on rendering of the diverse data/props in this function. 

In PersonalSata.test.js (input fields) i do not see any problems, but it shows an errors by getByTextId ("form") field. When i delete this line with the testing of form, the error goes to the next line and shows the same at the  getByTextId ("name"). 

In Cards.test.id (renders of array) in the line "expect(screen.getByText(dataItems.title)).toBeInTheDocument()" where the error occurs i suppose that (screen.getByText) ist wrong method for (dataItems.title). Or the whole construction can be written wrong!

Please see the last two commits in my pull request: https://github.com/nazardiachok/Capstone-New-Project/pull/22
and some pictures where the errors is to be seen:
<img width="1440" alt="Screenshot 2022-11-12 at 13 51 13" src="https://user-images.githubusercontent.com/111637550/201476612-a053f135-8524-4102-aad5-df50b94162f1.png">
<img width="1440" alt="Screenshot 2022-11-12 at 13 51 25" src="https://user-images.githubusercontent.com/111637550/201476627-01c511d9-28f6-4906-84b4-f14bf679b88d.png">
<img width="1440" alt="Screenshot 2022-11-12 at 14 03 30" src="https://user-images.githubusercontent.com/111637550/201476635-b3699c7e-3ead-44a8-a8a3-cd0165c64655.png">

Thanks in advance!!!

**Comment by VladF146** [comment, marked as answer]
**PersonalData.js**
Hello Nazar, if you want to "**getByTestId**", your component needs a data-testid attribute: `<form data-testid="form">`
Or you could add an aria label to your form to be able to use "**getByRole**": `<form aria-label="form">`

In practice you should avoid using testId, if you have other options. For your name and email inputs this would be preferred:

```javascript
...
const emailInput = screen.getByRole("textbox", {name: /email/i});
const nameInput = screen.getByRole("textbox", {name: /name/i});
...
```

**Card.js**
You could simplify you test by just checking, whether the count of the matching items is the same as the length of the props-Array:

```javascript
...
const cardHeadings = screen.getAllByRole("heading", {level: 5});
expect(cardHeadings.length).toBe(dataItems.length);
...
```

**Reply by nazardiachok** [reply]
Hello Vlad, thank you very much for your help!!! It works!!! But in Card.js in this code ... expect(cardHeadings.length).**toBe**(dataItems.length) ... .toBe() does not work. Also .toEqual and .toStrictEqual does not work. But what works hier is .not.toBe(). Thank you !!!

**Reply by VladF146** [reply]
It doesn‘t work because you use `h5` for your title and another `h5` for amount. That is why you have twice as many `h5` as dataItems. Just use a `div` for the amount, price instead of the second `h5`.

**Reply by nazardiachok** [reply]
thanks!!!
